### PR TITLE
Allow fmf-id to be exported from run discover

### PR DESCRIPTION
**Pull request by dkarpele**
#880

Add an option `fmf-id` to discover step. It allows returning fmf identifiers for tests discovered in plan.
Catch the exception found during implementation. In master we have:
`$ tmt run test --name=/lint/tests plan --name=/plans/sanity/lint discover --fmf-id`
```
Traceback (most recent call last):
...
  File "/home/dkarpele/fork/tmt/tmt/base.py", line 169, in fmf_id
    '/', os.path.relpath(fmf_root, git_root))
  File "/usr/lib64/python3.9/posixpath.py", line 454, in relpath
    raise ValueError("no path specified")
```
because `self.node.root = None` in https://github.com/psss/tmt/blob/67dae23a77c4b5f0f9ab0d71470e343041ef4454/tmt/base.py#L166
I don't know if it's expected or not, maybe weird case. I'm returning git root path in this case.
Tests were added.

### CODE REVIEW

**Comment by dkarpele** [comment]
> Hi @dkarpele, path output is wrong (for the intended usage == pipe fmf-id to other tool for scheduling)

It wasn't easy to find who makes the path wrong. It updates in `DiscoverFmf` plugin. The only way I found to make path the same as in `test export --fmf-id` is to use `fmf_root` as `tree_path`

**Comment by dkarpele** [comment]
@lukaszachy  There is one more open question. 
```
[dkarpele@redhat tmt]$ tmt run -r discover --how fmf --fmf-id tests --name /tests/unit
---
name: /tests/unit
url: https://github.com/dkarpele/tmt.git
ref: dkarpele-880
---
name: /tests/unit
url: https://github.com/dkarpele/tmt.git
ref: dkarpele-880
```

It reports the same fmf-ids twice because test `/tests/unit` exists in 2 different plans. I think we should avoid it. My suggestion is to add tests responsible to already printed fmf-ids to the file. After that on the next plan, check that tests exists in the file. If tests exist in the file => print nothing.
What do you think?

**Comment by lukaszachy** [comment]
@dkarpele Could be more user friendly for sure. It just needs to make sure to check for whole fmf-id not just name (ref is allowed to differ between plans).

**Comment by dkarpele** [comment]
Thanks for the review @lukaszachy. All comments were addressed.

**Comment by lukaszachy** [comment]
Hm, following plan doesn't work for me (plan is placed in fresh directory with just `tmt init`)
```
execute:
    how: tmt
/first:
    discover:
        how: fmf
        url: https://github.com/psss/fmf.git
```
Even `tmt run discover -h fmf --fmf-id --url https://github.com/psss/fmf.git` doesn't print anything.
Run without `fmf-id` discover prints that  "summary: 8 tests selected".

**Comment by lukaszachy** [comment]
It worked before 49dc591. Seems there are different expectations for 'local' discover (same as tmt t export) and 'remote' one using `url` attribute where tmt t is not possible.

**Comment by dkarpele** [comment]
@lukaszachy 
1. I think we expect `url: https://github.com/psss/fmf.git` in that case. 
2. Need to take into account `--ref` option also. If `--ref` wasn't provided then `master` should be used.
3. And it's unclear for me which `path` should be in that case.
[Path to the metadata tree root. Should be relative to the git repository root if url provided, absolute local filesystem path otherwise. ](https://fmf.readthedocs.io/en/latest/concept.html#identifiers)

If `url` is provided metadata Clones from 'https://github.com/psss/fmf.git' to dir like '/var/tmp/tmt/run-066/plans/sanity/lint/discover/default/tests'
The `workdir: /var/tmp/tmt/run-066/plans/sanity/lint/discover`
Are we looking to the path from current dir to '/var/tmp/tmt/run-066/plans/sanity/lint/discover/default/tests'?

**Comment by lukaszachy** [comment]
Hm, missing `--ref` isn't necessary master: It is default branch of the repo so best would be to omit it from the output.

The use case for this option is to pipe fmf-ids from tmt to other tool (e.g. beaker workflow), similar to 'tmt tests export --fmf-id'. Difference between these commands is that `tmt tests` is always local but `tmt run discover` can fetch from remotes (and soon from source rpm). However the output should be self contained and usable on another host.

With this in mind: path to the local filesystem is not usable outside of the host. 
Same if url is missing (in both cases you cannot use the output to schedule tests on the remote) 
Maybe raise an exception in those cases?

**Comment by dkarpele** [comment]
@lukaszachy,
if I got your point: using `--url` option we can't show correct `path` (`os.path.relpath(fmf_root, git_root`), because`fmf_root` is in workdir, not in dir with `.git`. That's why it doesn't make sense to show any output for `tmt run discover --fmf-id --url="url"`. And raise an exception like "--url can't be used together with --fmf-id"?

**Comment by lukaszachy** [comment]
@dkarpele Exact opposite. Local-only paths cannot be used with fmf-id.

Think about this use case:
I have plan which uses tests from external repo (e.g. shared tests namespace) and I want to schedule (just tests, no prepare/finish steps) in e.g. Beaker. So I need to `tmt run discover --fmf-id | <scheduling_tool>`
Since external system doesn't access to have any other data than fmf-ids they need to specify valid test objects.

**Comment by dkarpele** [comment]
reminder:
2 cases should be implemented:
1. Without `--url` option. `fmf_root` and `git_root` are discoverable (.git and .fmf are in one project). In this case `git_root` should be in testdir (fmf.py > DiscoverFmf#L129). `fmf_root` should be in workdir.
2. With `--url` option. `fmf_root` should be in workdir. `git_root` should be exported from url

**Comment by lukaszachy** [comment]
Hm, there is #941 which will allow to duplicate tests. IMO `discover --fmf-id` should be consistent but that means tmt should allow to print duplicate fmf-ids within same plan.
How about saying that if you specify more plans you get all tests they discover, including duplicates? Basically meaning that output from https://github.com/psss/tmt/pull/932#issuecomment-956724963 will be correct behaviour.

**Comment by dkarpele** [comment]
I think in #941 and #950 there is a discussion about test duplication **within the same plan**
One clarification: In https://github.com/psss/tmt/pull/932#issuecomment-956724963 I describe the case with test duplication **in different plans**. That's why I still think that in https://github.com/psss/tmt/pull/932#issuecomment-956724963 we should avoid fmf-id duplcation. @lukaszachy what do you think?

**Comment by lukaszachy** [comment]
Frankly, I'm not sure about cost/benefit ratio when avoiding duplicates because from the fmf-id output one cannot tell boundaries between plans.

**Comment by dkarpele** [comment]
Hi @lukaszachy, I've prepared a patch for cases described in the table. The test cases are described in `filtering.sh`. More tests will be prepared soon. I commented the sections that was responsible for removing duplicates as this feature was considered useless (the same for the test). Please have a look when you are ready.

File: tmt/steps/discover/__init__.py [hunk 1]
```diff
@@ -192,7 +231,33 @@ def base_command(cls, method_class=None, usage=None):
         @click.option(
             '-h', '--how', metavar='METHOD',
             help='Use specified method to discover tests.')
+        @click.option(
+            '--fmf-id', default=False, is_flag=True,
+            help='Show fmf identifiers for tests discovered in plan.')
         def discover(context, **kwargs):
+            if kwargs.get('fmf_id'):
```
Comment (lukaszachy) [inline tmt/steps/discover/__init__.py:238]: Shouldn't whole fmf-id functionality be moved to `DiscoverFmf` class?

File: tmt/steps/discover/__init__.py [hunk 2]
```diff
@@ -192,7 +231,33 @@ def base_command(cls, method_class=None, usage=None):
         @click.option(
             '-h', '--how', metavar='METHOD',
             help='Use specified method to discover tests.')
+        @click.option(
+            '--fmf-id', default=False, is_flag=True,
```
Comment (lukaszachy) [inline tmt/steps/discover/__init__.py:235]: I think this should be just in `DiscoverFmf`. IMO there is no point to have common option for all discover plugins when it can be used only with `how:fmf`

File: tmt/steps/discover/__init__.py [hunk 3]
```diff
@@ -192,7 +231,33 @@ def base_command(cls, method_class=None, usage=None):
...
+                    raise tmt.utils.GeneralError(
+                        f"`tmt run discover --fmf-id` is supported only"
+                        f" for `fmf` method.")
+                import subprocess
+                context.parent.params['quiet'] = True
+                context.parent.params['debug'] = 0
+                context.parent.params['verbose'] = 0
+                # Raise an exception if --fmf-id uses w/o --url and git root
+                # doesn't exist for ALL discovered plans
+                url = tmt.utils.Common.get_fmf_attr('discover', 'url')
```
Comment (lukaszachy) [inline tmt/steps/discover/__init__.py:249]: this shouldn't be necessary when run in `DiscoverFmf`'s go() - it should take care merging cmdline/plan options (opt() call)

**Review by lukaszachy — CHANGES_REQUESTED** [review]
Feature and test coverage wise looks good.
IMO functionality should be moved to the DiscoverFmf plugin itself - and newly added get_fmf_attr or steps might not be necessary

File: tmt/steps/discover/__init__.py (hunk 1 above)
Comment (dkarpele) [reply tmt/steps/discover/__init__.py:232]: Thanks for the review Lukas.
I've moved the option `--fmf-id` to `DiscoverFmf` and it helped avoiding the checking of `-h shell --fmf-id`. The `click.Context` module checks it.
Unfortunately it's not possible to move the condition above to `DiscoverFmf` because disabling logging (debug, verbose) should be done immediately after start. In `DiscoverFmf.go()` it's "too late" because some data will be already printed.

File: tmt/steps/discover/__init__.py (hunk 3 above)
Comment (dkarpele) [reply tmt/steps/discover/__init__.py:240]: I've tried to move that to `DiscoverFmf.go()` and it works fine in all cases except one:
"fmf-id (w/o and w/ url): non-git directory": If in non-git directory exists plan w/ and w/o url, then only fmf-ids for the plan w/ url should be reported.
Without `get_fmf_attr` the exception will be raised when non-url plan will be found and the whole output will be mess. It will look like the list of fmf-ids with `tmt.utils.GeneralError` at the bottom. Or just `GeneralError` at the top.
I can still move the whole piece of code to `DiscoverFmf.go()` but **with** `get_fmf_attr`

Comment (dkarpele) [reply tmt/steps/discover/__init__.py]: UPD: I've decided to move the code above **with** `get_fmf_attr` to `DiscoverFmf.go()`. It makes sense. All tests passed except fedora-rawhide: there is an infrastructure issue.

File: tmt/base.py [hunk 4]
```diff
@@ -143,14 +143,17 @@ def fmf_id(self):
 
         def run(command):
             """ Run command, return output """
+            cwd = fmf_root if self.steps() == {'discover'} else None
```
Comment (lukaszachy) [inline tmt/base.py:146]: Q: Is there a situation where setting cwd to fmf_root will break stuff?
I don't like `self.steps() == {'discover'} ` - it is not easy to see why magic value is `{'discover'}` and imho this check is not necessary

File: tmt/base.py [hunk 5]
```diff
@@ -159,16 +162,16 @@ def run(command):
             remote_name = 'origin'
         remote = run(f"git config --get remote.{remote_name}.url")
         fmf_id['url'] = tmt.utils.public_git_url(remote)
+        if fmf_id['url'] == '' and self.steps() == {'discover'}:
```
Comment (lukaszachy) [inline tmt/base.py:165]: Again - magic constant `{'discover'}` IMHO empty 'url' could be checked in fmf-id caller (as it matter only there)

File: tmt/steps/discover/__init__.py [hunk 6]
```diff
@@ -146,6 +147,40 @@ def go(self):
...
                 self._tests.append(test)
 
+        # Show fmf identifiers for tests discovered in plan
+        if self.opt('fmf_id'):
+            # don't run steps except discover
+            self._context.obj.steps = {'discover'}
+            if self.tests():
+                fmf_id_list = [tmt.utils.dict_to_yaml(test.fmf_id, start=True)
+                               for test in self.tests()
+                               if test.fmf_id]
```
Comment (lukaszachy) [inline tmt/steps/discover/__init__.py:157]: IMHO here should be check that url is set (`if  'url' in test.fmf_id`)

File: tmt/steps/discover/fmf.py [hunk 7]
```diff
@@ -147,6 +151,21 @@ def go(self):
         testdir = os.path.join(self.workdir, 'tests')
         dist_git_source = self.get('dist-git-source', False)
 
+        # Raise an exception if --fmf-id uses w/o --url and git root
+        # doesn't exist for ALL discovered plans
+        if self.opt('fmf_id') and \
+                not tmt.utils.Common.get_fmf_attr('discover', 'url'):
```
Comment (lukaszachy) [inline tmt/steps/discover/fmf.py:157]: url might be set in plan or CLI and `url = self.opt('url')` has already taken care getting the correct value
```suggestion
                not url:
```

File: tmt/utils.py [hunk 8]
```diff
@@ -184,6 +184,13 @@ def _fmf_context(self):
         except AttributeError:
             return dict()
 
+    def steps(self):
```
Comment (lukaszachy) [inline tmt/utils.py:187]: IMHO this method is not necessary (and doen't make sense for Test object). Is the only usage for fmf-id property?

File: tmt/utils.py [hunk 9]
```diff
@@ -532,11 +539,29 @@ def workdir(self):
             create_directory(self._workdir, 'workdir', quiet=True)
         return self._workdir
 
+    @staticmethod
```
Comment (lukaszachy) [inline tmt/utils.py:542]: IMO not necessary - cli/plan merging might happen and tmt should take care of it.

File: tests/discover/filtering.sh [hunk 10]
```diff
@@ -45,6 +45,170 @@ rlJournalStart
...
+
+        tests_amount=$(echo $tests_list | wc -w)
+        rlAssertEquals "Check that number of fmf-ids equals to tests number" \
+                       "$ids_amount" "$tests_amount"
+        rlAssertEquals "Check url" "$url_discover" "$url_fmf_id"
+        rlRun "rm -rf $tmp_dir1 $tmp_dir2"
+    rlPhaseEnd
+
+    # Raise an exception if current dir doesn't have .fmf
+    rlPhaseStartTest "fmf-id (w/o url): current dir doesn't have fmf metadata"
```
Comment (lukaszachy) [inline tests/discover/filtering.sh:203]: I experimented a bit with my proposal fmf-id /url changes and behaviour changed to raise  `--fmf-id` without `url` option can be used only within git repo.` instead which is imho more appropriate (one can specify plan using tmt options)

File: tests/discover/filtering.sh [hunk 11]
```diff
@@ -45,6 +45,170 @@ rlJournalStart
...
+        rlRun "tmt init --template base"
+        rlRun "tmt run -rdvvv discover -h fmf --fmf-id finish 2>&1 \
+               | tee output" 2
+        rlAssertGrep "\`tmt run discover --fmf-id\` without \`url\` option \
+can be used only within git repo." output
+        rlRun "rm -rf $tmp_dir"
+    rlPhaseEnd
+
+    # If in non-git directory exist plan w/ and w/o url, then only fmf-ids for
+    # the plan w/ url should be reported
```
Comment (lukaszachy) [inline tests/discover/filtering.sh:170]: Is this assumption correct? If I have one plan which cannot report fmf-id if called in its own, why this error should be hidden if it is called with another but working plan? --fmf-id is expected for all plans not only those which can do it.

**Review by lukaszachy — CHANGES_REQUESTED** [review]
I apology for late review. Few comments. 
Overall it looks great and very useful

File: tmt/base.py (hunk 4 above)
Comment (dkarpele) [reply tmt/base.py:146]: This code is responsible for the case mentioned by you below https://github.com/teemtee/tmt/pull/932#discussion_r821672134

File: tmt/base.py (hunk 5 above)
Comment (dkarpele) [reply tmt/base.py:165]: (same) This code is responsible for the case mentioned by you below https://github.com/teemtee/tmt/pull/932#discussion_r821672134

File: tmt/steps/discover/fmf.py (hunk 7 above)
Comment (dkarpele) [reply tmt/steps/discover/fmf.py:157]: (same) This code is responsible for the case mentioned by you below https://github.com/teemtee/tmt/pull/932#discussion_r821672134

File: tmt/base.py (hunk 4 above)
Comment (lukaszachy) [reply tmt/base.py:146]: I know, but is there use case where cwd can't be set to fmf_root ? From what I tried cwd can be safely set to fmf_root each time and thus such check is not necessary

File: tests/discover/filtering.sh (hunk 11 above)
Comment (dkarpele) [reply tests/discover/filtering.sh:170]: Thanks for the review @lukaszachy, in this case,
1) If we report an error
`tmt run discover --fmf-id without url option can be used only within git repo.` it will be wrong for "working plan" as it has url. 
2) If we report an error then we shouldn't show fmf-ids for "working plan" also. Otherwise we will have a case:
```
(venv) [dkarpele@redhat plans]$ ls
example.fmf  non-url.fmf
(venv) [dkarpele@redhat plans]$tmt run -rv discover -h fmf --fmf-id finish
---
name: /tests/clean/guests
url: https://github.com/teemtee/tmt
ref: main
...
`tmt run discover --fmf-id` without `url` option can be used only within git repo.
```

It breaks the original idea of beaker workflow. `tmt run plan --fmf-id | bkr workflow-tomorrow f-34 x86_64 --fmf-id -`. We can't pass error message to bkr.
OR if non url file goes alphabetically first.
```
(venv) [dkarpele@redhat plans]$ ls
a-non-url.fmf  example.fmf
(venv) [dkarpele@redhat plans]$ tmt run -rdddvvv discover -h fmf --fmf-id finish
`tmt run discover --fmf-id` without `url` option can be used only within git repo.
```
It breaks the original idea of beaker workflow. `tmt run plan --fmf-id | bkr workflow-tomorrow f-34 x86_64 --fmf-id -`. We can't pass error message to bkr.

File: tmt/base.py (hunk 5 above)
Comment (lukaszachy) [reply tmt/base.py:165]: IMHO 'self.steps()' is not necessary and such method doesn't make sense for Test objects. Empty url is problem only for `discover -h fmf --fmf-id` usage and as such should be handled there (e.g. `'url' in fmf_id` check)

File: tmt/steps/discover/fmf.py (hunk 7 above)
Comment (lukaszachy) [reply tmt/steps/discover/fmf.py:157]: There might be difference between fmf attributes and final value (cli preferences, future defaults etc.) IMO 'url' is already set correctly

Is there use case where `url` is different from `get_fmf_attr('discover', 'url')` and the later has expected value?

Comment (dkarpele) [reply tmt/steps/discover/fmf.py:157]: With `non-url.fmf` and `url.fmf` in the same run.
```
(venv) [dkarpele@redhat plans]$ cat non-url.fmf 
summary: Basic smoke test
discover:
    how: fmf
execute:
    how: tmt
$ tmt run -rdddvvv discover -h fmf --fmf-id finish
``` 
If I stop debugger on this line then `url = None` and `tmt.utils.Common.get_fmf_attr('discover', 'url') = 'https://github.com/teemtee/tmt'`

But if we agree "that user should fix their command line to select plans which can success for 'discover --fmf-id'" then all this condition isn't necessary.

Comment (dkarpele) [reply tmt/steps/discover/fmf.py:157]: But I need to know in advance that one of the plans doesn't have url before printing the output for working plan...

Comment (lukaszachy) [reply tmt/steps/discover/fmf.py:157]: Ah, I didn't see this use case

> But if we agree "that user should fix their command line to select plans which can success for 'discover --fmf-id'" then this condition isn't necessary.

Yes, please. Failing is IMO better than to hide tests which don't have url set.

File: tests/discover/filtering.sh (hunk 10 above)
Comment (dkarpele) [reply tests/discover/filtering.sh:206]: This test covers only an exception when fmf.Tree can't be initialized. fmf.py#175

File: tmt/steps/discover/fmf.py [hunk 12]
```diff
@@ -147,6 +152,43 @@ def go(self):
         testdir = os.path.join(self.workdir, 'tests')
         dist_git_source = self.get('dist-git-source', False)
 
+        # Raise an exception if --fmf-id uses w/o url and git root
+        # doesn't exist for discovered plan
+        if self.opt('fmf_id'):
+            def print_error(plan_name=None):
```
Comment (lukaszachy) [inline tmt/steps/discover/fmf.py:158]: Name is a bit misleading - It's doesn't print error at all, only raises exception.
How about 'assert_git_url' ? (or something among those lines).

File: tmt/steps/discover/fmf.py [hunk 13]
```diff
@@ -147,6 +152,43 @@ def go(self):
...
+                        check=True)
+                except subprocess.CalledProcessError:
+                    raise tmt.utils.GeneralError(
+                        f"`tmt run discover --fmf-id` without `url` option in "
+                        f"plan `{plan_name}` can be used only within"
+                        f" git repo.")
+
+            try:
+                fmf_tree = fmf.Tree(os.getcwd())
+            except fmf.utils.RootError:
```
Comment (lukaszachy) [inline tmt/steps/discover/fmf.py:173]: One is unable to create cmdline only plan, e.g. `tmt run -r  discover -h fmf --fmf-id --url  URL` in directory without fmf root.
IMO not existing fmf_tree is valid use case, so later `enumerate(fmf_tree.climb()` can be part of try/except and exception will be ignored.
If there is no url on CLI it will be covered by `not url` check.

Comment (lukaszachy) [reply tmt/steps/discover/fmf.py:173]: This use case might to be that silly - e.g. ad-hoc scheduling from test repo without cloning it locally first.

Comment (dkarpele) [reply tmt/steps/discover/fmf.py:173]: Thanks Lukas, very good point. This is a valid case but I don't understand how to fix that. Could you please provide a patch to ignore exception?

Another point. I'm thinking that there is bug if we call `tmt run -r discover -h fmf` in directory without fmf root: https://cpaste.org/?0b98e395e9727532#3RGXsT1eu8fc8b3uzXZWYbk8KxGPDHNTEpN7VPQd97XY

Comment (lukaszachy) [reply tmt/steps/discover/fmf.py]: In all cases I'd raise `tmt.utils.DiscoverError` as discover plugin can't continue and that way tmt finishes with proper message and exit code - https://cpaste.org/?d2bb05249f808666#GKi6Ao1FMibsGa5Jjy4t4WN95bTmjNGgrZVuap3aYaPS

**Comment by dkarpele** [comment]
@lukaszachy thanks for the approval!
I'd like to leave this to you. I think you will describe the commit better.
